Libraries/LibGemini/Job.cpp

```diff
@@ -0,0 +1,161 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibGemini/GeminiResponse.h>
+#include <LibGemini/Job.h>
+#include <stdio.h>
+#include <unistd.h>
+
+//#define JOB_DEBUG
+
+namespace Gemini {
+
+Job::Job(const GeminiRequest& request)
+    : m_request(request)
+{
+}
+
+Job::~Job()
+{
+}
+
+void Job::on_socket_connected()
+{
+    register_on_ready_to_write([this] {
+        if (m_sent_data)
+            return;
+        m_sent_data = true;
+        auto raw_request = m_request.to_raw_request();
+#ifdef JOB_DEBUG
+        dbg() << "Job: raw_request:";
+        dbg() << String::copy(raw_request).characters();
+#endif
+        bool success = write(raw_request);
+        if (!success)
+            deferred_invoke([this](auto&) { did_fail(Core::NetworkJob::Error::TransmissionFailed); });
+    });
+    register_on_ready_to_read([this] {
+        if (is_cancelled())
+            return;
+
+        if (m_state == State::InStatus) {
+            if (!can_read_line())
+                return;
+
+            auto line = read_line(PAGE_SIZE);
+            if (line.is_null()) {
+                fprintf(stderr, "Job: Expected status line\n");
+                return deferred_invoke([this](auto&) { did_fail(Core::NetworkJob::Error::TransmissionFailed); });
+            }
+
+            auto parts = String::copy(line, Chomp).split_limit(' ', 2);
+            if (parts.size() != 2) {
+                fprintf(stderr, "Job: Expected 2-part status line, got '%s'\n", line.data());
+                return deferred_invoke([this](auto&) { did_fail(Core::NetworkJob::Error::ProtocolFailed); });
+            }
+
+            bool ok;
+            m_status = parts[0].to_uint(ok);
+            if (!ok) {
+                fprintf(stderr, "Job: Expected numeric status code\n");
+                return deferred_invoke([this](auto&) { did_fail(Core::NetworkJob::Error::ProtocolFailed); });
+            }
+
+            m_meta = parts[1];
+
+            if (m_status >= 10 && m_status < 20) {
+                m_state = State::Finished;
+            } else if (m_status >= 20 && m_status < 30) {
+                m_state = State::InBody;
+            } else if (m_status >= 30 && m_status < 40) {
+                m_state = State::Finished;
+            } else if (m_status >= 40 && m_status < 50) {
+                m_state = State::Finished;
+            } else if (m_status >= 50 && m_status < 60) {
+                m_state = State::Finished;
+            } else if (m_status >= 60 && m_status < 70) {
+                m_state = State::InBody;
+            } else {
+                fprintf(stderr, "Job: Expected status between 10 and 69; instead got %d\n", m_status);
+                return deferred_invoke([this](auto&) { did_fail(Core::NetworkJob::Error::ProtocolFailed); });
+            }
+
+            return;
+        }
+
+        ASSERT(m_state == State::InBody || m_state == State::Finished);
+        ASSERT(can_read());
```

Comment:
What if the remote closes the connection at this point?

Moreover, does `can_read()` return `true` if you could read more, but that would block?

Reply:
Actually this is totally broken if there's no body - we should only be checking for `can_read()` when we're expecting a response body. Good catch!

Reply:
That's not what I meant actually (good catch yourself!)

I mean if the remote abruptly closes the connection when we expect the body, we should return a failure rather than aborting.

Reply:
Ah! Yes, I see what you mean now. Two bugs for the price of one!

Reply:
I've decided the right thing to do here is just to delete the `can_read()` check completely - a zero-length body is totally valid, and results in EOF directly after sending the headers.